src/commonMain/kotlin/exchange.dydx.abacus/calculator/V2/TradeInput/TradeInputMarketOrderCalculator.kt

```diff
@@ -0,0 +1,503 @@
+        stepSize: Double,
+    ): TradeInputMarketOrder? {
+        return if (usdcSize != null && usdcSize != Numeric.double.ZERO) {
+            if (orderbook != null) {
+                val desiredUsdcSize = usdcSize
+                var sizeTotal = Numeric.double.ZERO
+                var usdcSizeTotal = Numeric.double.ZERO
+                var worstPrice: Double? = null
+                var filled = false
+                val marketOrderOrderBook = mutableListOf<InternalOrderbookTick>()
+
+                orderbookLoop@ for (element in orderbook) {
+                    val entryPrice = element.price
+                    val entrySize = element.size
+
+                    if (entryPrice > Numeric.double.ZERO) {
+                        val entryUsdcSize = entrySize * entryPrice
+                        filled = (usdcSizeTotal + entryUsdcSize >= desiredUsdcSize)
+
+                        var matchedSize = entrySize
+                        var matchedUsdcSize = entryUsdcSize
+                        if (filled) {
+                            matchedUsdcSize = desiredUsdcSize - usdcSizeTotal
+                            matchedSize = matchedUsdcSize / entryPrice
+                            matchedSize =
+                                Rounder.quickRound(
+                                    matchedSize,
+                                    stepSize,
+                                )
+                            matchedUsdcSize = matchedSize * entryPrice
+                        }
+                        sizeTotal += matchedSize
+                        usdcSizeTotal += matchedUsdcSize
+
+                        worstPrice = entryPrice
+                        marketOrderOrderBook.add(matchingOrderbookEntry(element, matchedSize))
+                        if (filled) {
+                            break@orderbookLoop
+                        }
+                    }
+                }
+                createMarketOrderWith(
+                    orderbook = marketOrderOrderBook,
+                    size = sizeTotal,
+                    usdcSize = usdcSizeTotal,
+                    worstPrice = worstPrice,
+                    filled = filled,
+                )
+            } else {
+                createMarketOrderWith(
+                    orderbook = listOf<InternalOrderbookTick>(),
+                    size = Numeric.double.ZERO,
+                    usdcSize = usdcSize,
+                    worstPrice = null,
+                    filled = false,
+                )
+            }
+        } else {
+            null
+        }
+    }
+
+    private fun createMarketOrderFromLeverage(
+        leverage: Double,
+        market: InternalMarketState?,
+        subaccount: InternalSubaccountState?,
+        user: InternalUserState?,
+    ): TradeInputMarketOrder? {
+        val stepSize = market?.perpetualMarket?.configs?.stepSize ?: 0.001
+        val equity = subaccount?.calculated?.get(CalculationPeriod.current)?.equity ?: return null
```

Comment:
I think @moo-onthelawn just made a bug fix regarding equity in this calculator function.

https://github.com/dydxprotocol/v4-abacus/commit/96f520878fb8f3f48bb4a083f5fa132dcc945bd1

Reply:
Yeah.. it's been added to the follow-up PR.